### PR TITLE
HDFS-16923. [SBN read] getlisting RPC to observer will throw NPE if path does not exist

### PR DESCRIPTION
**Pull request by ZanderXu**
### Description of PR
Jira: [HDFS-16923](https://issues.apache.org/jira/browse/HDFS-16923)

The getlisting rpc will throw NPE if the path does not exist.

### CODE REVIEW

**Comment by ZanderXu** [comment]
@zhengchenyu @xkrogen Master, please help me to review this NPE imported by HDFS-16732

File: hadoop-hdfs-project/hadoop-hdfs-rbf/src/test/java/org/apache/hadoop/hdfs/server/federation/router/TestObserverWithRouter.java [hunk 1]
```diff
@@ -146,6 +148,19 @@ public void testObserverRead() throws Exception {
     internalTestObserverRead();
   }
 
+  @Test
+  public void testGetListingWithObserverRead() throws Exception {
```
Comment (xkrogen) [inline hadoop-hdfs-project/hadoop-hdfs-rbf/src/test/java/org/apache/hadoop/hdfs/server/federation/router/TestObserverWithRouter.java]: Feels to me that this test probably belongs in `TestObserverNode`? It's not really related to Router/federation.

**Review by xkrogen — COMMENTED** [review]
Thanks @ZanderXu !

I re-reviewed HDFS-16732 and I think we also need to do the same thing for the block added in `getFileInfo()` (null guard on `stat`), WDYT?

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 37s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 1 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |  15m 11s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  31m 55s |  |  trunk passed  |
| +1 :green_heart: |  compile  |   6m 18s |  |  trunk passed with JDK Ubuntu-11.0.17+8-post-Ubuntu-1ubuntu220.04  |
| +1 :green_heart: |  compile  |   6m 12s |  |  trunk passed with JDK Private Build-1.8.0_352-8u352-ga-1~20.04-b08  |
| +1 :green_heart: |  checkstyle  |   1m 15s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   2m 12s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   1m 46s |  |  trunk passed with JDK Ubuntu-11.0.17+8-post-Ubuntu-1ubuntu220.04  |
| +1 :green_heart: |  javadoc  |   2m 25s |  |  trunk passed with JDK Private Build-1.8.0_352-8u352-ga-1~20.04-b08  |
| +1 :green_heart: |  spotbugs  |   4m 44s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  23m 58s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 27s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   2m  5s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   6m 26s |  |  the patch passed with JDK Ubuntu-11.0.17+8-post-Ubuntu-1ubuntu220.04  |
| +1 :green_heart: |  javac  |   6m 26s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   5m 49s |  |  the patch passed with JDK Private Build-1.8.0_352-8u352-ga-1~20.04-b08  |
| +1 :green_heart: |  javac  |   5m 49s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   1m  4s |  |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   1m 57s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   1m 25s |  |  the patch passed with JDK Ubuntu-11.0.17+8-post-Ubuntu-1ubuntu220.04  |
| +1 :green_heart: |  javadoc  |   2m 18s |  |  the patch passed with JDK Private Build-1.8.0_352-8u352-ga-1~20.04-b08  |
| +1 :green_heart: |  spotbugs  |   4m 47s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  24m 27s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| -1 :x: |  unit  | 205m 53s | [/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5400/1/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt) |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  unit  |  20m 34s |  |  hadoop-hdfs-rbf in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 48s |  |  The patch does not generate ASF License warnings.  |
|  |   | 376m 52s |  |  |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.server.namenode.TestAuditLogger |
|   | hadoop.hdfs.server.namenode.TestFsck |
|   | hadoop.hdfs.server.namenode.TestAuditLogs |
|   | hadoop.hdfs.server.namenode.TestFSNamesystemLockReport |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.42 ServerAPI=1.42 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5400/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/5400 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets |
| uname | Linux 0504b05c1ae8 4.15.0-200-generic #211-Ubuntu SMP Thu Nov 24 18:16:04 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / deaaebdf677274d3c990e76d6904d93ef3fbdfa9 |
| Default Java | Private Build-1.8.0_352-8u352-ga-1~20.04-b08 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.17+8-post-Ubuntu-1ubuntu220.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_352-8u352-ga-1~20.04-b08 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5400/1/testReport/ |
| Max. process+thread count | 3163 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs hadoop-hdfs-project/hadoop-hdfs-rbf U: hadoop-hdfs-project |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5400/1/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by ZanderXu** [comment]
@simbadzina Master, sorry for the late response. 

>  I think we also need to do the same thing for the block added in getFileInfo() (null guard on stat), WDYT?

`stat instanceof HdfsLocatedFileStatus` will return false if the stat is null, so there is no bugs in getFileInfo().

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 38s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  1s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  1s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 1 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  38m 55s |  |  trunk passed  |
| +1 :green_heart: |  compile  |   1m 27s |  |  trunk passed with JDK Ubuntu-11.0.17+8-post-Ubuntu-1ubuntu220.04  |
| +1 :green_heart: |  compile  |   1m 24s |  |  trunk passed with JDK Private Build-1.8.0_352-8u352-ga-1~20.04-b08  |
| +1 :green_heart: |  checkstyle  |   1m  7s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 31s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   1m  8s |  |  trunk passed with JDK Ubuntu-11.0.17+8-post-Ubuntu-1ubuntu220.04  |
| +1 :green_heart: |  javadoc  |   1m 36s |  |  trunk passed with JDK Private Build-1.8.0_352-8u352-ga-1~20.04-b08  |
| +1 :green_heart: |  spotbugs  |   3m 26s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  22m 46s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   1m 17s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 19s |  |  the patch passed with JDK Ubuntu-11.0.17+8-post-Ubuntu-1ubuntu220.04  |
| +1 :green_heart: |  javac  |   1m 19s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 13s |  |  the patch passed with JDK Private Build-1.8.0_352-8u352-ga-1~20.04-b08  |
| +1 :green_heart: |  javac  |   1m 13s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   0m 52s |  |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   1m 20s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   0m 51s |  |  the patch passed with JDK Ubuntu-11.0.17+8-post-Ubuntu-1ubuntu220.04  |
| +1 :green_heart: |  javadoc  |   1m 27s |  |  the patch passed with JDK Private Build-1.8.0_352-8u352-ga-1~20.04-b08  |
| +1 :green_heart: |  spotbugs  |   3m 17s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  22m  6s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| -1 :x: |  unit  | 206m 40s | [/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5400/2/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt) |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 50s |  |  The patch does not generate ASF License warnings.  |
|  |   | 313m  6s |  |  |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.server.datanode.TestDirectoryScanner |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.42 ServerAPI=1.42 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5400/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/5400 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets |
| uname | Linux 2d169627bf6c 4.15.0-200-generic #211-Ubuntu SMP Thu Nov 24 18:16:04 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / c318352a45076e87e1aef2ef8278a176e5cb7335 |
| Default Java | Private Build-1.8.0_352-8u352-ga-1~20.04-b08 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.17+8-post-Ubuntu-1ubuntu220.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_352-8u352-ga-1~20.04-b08 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5400/2/testReport/ |
| Max. process+thread count | 3130 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-5400/2/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by xkrogen** [comment]
The only test failures is `TestDirectoryScanner.testThrottling`:
```
[ERROR] Tests run: 13, Failures: 3, Errors: 0, Skipped: 0, Time elapsed: 575.031 s <<< FAILURE! - in org.apache.hadoop.hdfs.server.datanode.TestDirectoryScanner
[ERROR] testThrottling(org.apache.hadoop.hdfs.server.datanode.TestDirectoryScanner)  Time elapsed: 162.513 s  <<< FAILURE!
java.lang.AssertionError: Throttle is too permissive
```
This one is a bit nondeterministic as it's actually running things and checking how long it takes. It doesn't look related.

Merging to `trunk` and `branch-3.3`. Thanks @ZanderXu ! I'm also going to see about getting this into 3.3.5 since we have [HDFS-16732](https://issues.apache.org/jira/browse/HDFS-16732) there and this is a pretty bad bug.

**Pull request merged**